kroxylicious/src/main/java/io/kroxylicious/proxy/internal/net/EndpointRegistry.java

```diff
@@ -238,6 +249,100 @@ public CompletionStage<Void> deregisterVirtualCluster(VirtualCluster virtualClus
+        }
+        else {
+            // set of nodeIds differ
+            return rec.getValue().thenCompose(u -> {
+                Map.Entry<Set<Integer>, CompletionStage<Void>> updated;
+                var cand = createReconcileEntry(nodeIds, new CompletableFuture<>());
+                if (vcr.reconciliationEntry().compareAndSet(rec, cand)) {
+                    // reconcile - work out which bindings are to be registered and which are to be removed.
+                    var bindingAddress = virtualCluster.getBindAddress().orElse(null);
+
+                    var toCreate = nodeIds.stream().map(i -> new VirtualClusterBrokerBinding(virtualCluster, i))
+                            .collect(Collectors.toCollection(ConcurrentHashMap::newKeySet));
+
+                    // first assemble the stream of de-registrations (and by side effect: update toCreate)
+                    var deregs = allOfStage(listeningChannels.values().stream()
+                            .filter(lcr -> lcr.unbindingStage.get() == null)
+                            .map(lcr -> lcr.bindingStage()
+                                    .thenCompose((acceptorChannel -> {
+                                        var bindings = acceptorChannel.attr(CHANNEL_BINDINGS);
+                                        if (bindings == null || bindings.get() == null) {
+                                            // nothing to do for this channel
+                                            return CompletableFuture.completedStage(null);
+                                        }
+                                        var bindingMap = bindings.get();
+
+                                        return allOfStage(bindingMap.values().stream()
+                                                .filter(vcb -> vcb.virtualCluster().equals(virtualCluster))
+                                                .filter(VirtualClusterBrokerBinding.class::isInstance)
+                                                .map(VirtualClusterBrokerBinding.class::cast)
+                                                .peek(toCreate::remove) // side effect
+                                                .filter(vcbb -> !nodeIds.contains(vcbb.nodeId()))
+                                                .map(vcbb -> deregisterBinding(virtualCluster, vcbb::equals)));
+                                    }))));
+
+                    // chain any binding registrations and organise for the reconciliations entry to complete
+                    var unused = deregs.thenCompose(u1 -> allOfStage(toCreate.stream()
+                            .map(vcbb -> {
+                                var brokerAddress = virtualCluster.getBrokerAddress(vcbb.nodeId());
+                                var endpoint = new Endpoint(bindingAddress, brokerAddress.port(),
+                                        virtualCluster.isUseTls());
+                                return registerBinding(endpoint, brokerAddress.host(), vcbb);
```

Comment:
I am quite possibly missing something. However I feel like a binding should know the endpoint. Otherwise what is it binding to?

Reply:
Originally, my idea was that the binding would provide the tuple (virtualcluster, brokerid) and we'd use the brokerid to lookup the upstream or downstream address.  In this PR, I've now included the upstreamTarget address in the binding.  I've left comments above above whether EndpointRegistry should take responsibility for downstream addresses too.
 Please see the comments on the subject above.